src/librustdoc/html/render/cache.rs

```diff
@@ -278,33 +253,84 @@ crate fn get_real_types<'tcx>(
     generics: &Generics,
     arg: &Type,
     tcx: TyCtxt<'tcx>,
-    recurse: i32,
-    res: &mut FxHashSet<(Type, ItemType)>,
-) -> usize {
-    fn insert(res: &mut FxHashSet<(Type, ItemType)>, tcx: TyCtxt<'_>, ty: Type) -> usize {
-        if let Some(kind) = ty.def_id().map(|did| tcx.def_kind(did).into()) {
-            res.insert((ty, kind));
-            1
+    recurse: usize,
+    res: &mut Vec<TypeWithKind>,
+) {
+    fn insert_ty(
+        res: &mut Vec<TypeWithKind>,
+        tcx: TyCtxt<'_>,
+        ty: Type,
+        mut generics: Vec<TypeWithKind>,
+    ) {
+        let is_full_generic = ty.is_full_generic();
+
+        if is_full_generic && generics.len() == 1 {
+            // In this case, no need to go through an intermediate state if the generics
+            // contains only one element.
+            //
+            // For example:
+            //
+            // fn foo<T: Display>(r: Option<T>) {}
+            //
+            // In this case, it would contain:
+            //
+            // ```
+            // [{
+            //     name: "option",
+            //     generics: [{
+            //         name: "",
+            //         generics: [
+            //             name: "Display",
+            //             generics: []
+            //         ]
+            //     }]
+            // }]
+            // ```
+            //
+            // After removing the intermediate (unnecessary) full generic, it'll become:
+            //
+            // ```
+            // [{
+            //     name: "option",
+            //     generics: [{
+            //         name: "Display",
+            //         generics: []
+            //     ]
+            // }]
+            // ```
+            //
+            // To be noted that it can work if there is ONLY ONE generic, otherwise we still
```

Comment:
Why does it only work if there's a single generic argument? It seems weird to have an outer object that just has `name: ""`

Reply:
It's what I tried to explain but I failed I guess haha.

So basically, arguments can only have a type and generics. What happens in case the argument is a full generic which is `Display + Debug + Write`? Then we can't say it's one type (from a search perspective) but a generic type (with empty name) having three generics.

Reply:
I just discussed this with Guillaume, and I think I can explain it now.

For purposes of search-indexing, `fn foo<T: Display>(x: T)` is treated as the type `fn(x: Display)`. This makes a search like `trait:Display ->` show `foo`. (It turns out you can filter by item type in rustdoc searches, which is part of what `ItemType` is for.)

However, `fn foo<T: Display + Clone>(x: T)`'s argument *cannot* be reduced into an `ItemType::Trait` because there are two traits involved, the `Display` trait and the `Clone` trait. So instead it's represented as a `ItemType::Generic`, which means there was more than one bound involved.

Reply:
Exactly. :)